### PR TITLE
Fix inheritance bug in CHC cex

### PR DESCRIPTION
**Pull request by leonardoalt**
The added tests used to throw because the wrong scope contract was taken into account,

### CODE REVIEW

**Comment by cameel** [comment]
So was this broken for inheritance in general? From your comments in the code I understand that the bug should affect code with virtual function calls but the tests cases don't have any virtuals and both the assertion and the variable are in the same contract.

**Comment by leonardoalt** [comment]
Not in general. It used to break when:
- the failed target is in a base contract
- the derived contract has extra state variables

File: test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+contract B is A {
+	uint a;
+	uint b;
+}
+
+contract C is B {
+	uint y;
+	uint z;
+	uint w;
```
Comment (leonardoalt) [inline test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol:18]: @cameel the bug was that the cex generator sees `C.f()` (which exists via inheritance) but uses only the state variables from `A` because the AST scope of `f` is `A`. So if the derived contracts did not have more state variables it was fine.

File: libsolidity/formal/Predicate.h [hunk 2]
```diff
@@ -166,6 +168,12 @@ class Predicate
 	/// nullptr if this predicate is not associated with a specific program AST node.
 	ASTNode const* m_node = nullptr;
 
+	/// The ContractDefinition that contains this predicate.
+	/// nullptr if this predicate is not associated with a specific contract.
+	/// This is unfortunately necessary because of virtual resolution for
+	/// function nodes.
+	ContractDefinition const* m_contractContext = nullptr;
```
Comment (hrkrshnn) [inline libsolidity/formal/Predicate.h:175]: Can this be `nullptr` currently?

**Review by hrkrshnn — COMMENTED** [review]
Looks good.

Are there tests with `super`?

```
contract A {
    function f() public virtual returns (uint256 r) {
        return 1;
    }
}


contract B is A {
    function f() public virtual override returns (uint256 r) {
        return super.f() | 2;
    }
}


contract C is A {
    function f() public virtual override returns (uint256 r) {
        return super.f() | 4;
    }
}


contract D is B, C {
    function f() public override(B, C) returns (uint256 r) {
        r = super.f() | 8;
        assert(r == 15);
    }
}
```

File: test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol (hunk 1 above)
Comment (cameel) [reply test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol:18]: So now, instead of getting the scope where function `f` is defined (which is contract `A`) it will take the contract it is currently analyzing (which is `C`)? Yeah, I see how this fixes things.

Does this also affect virtual calls in any way? I.e. if contract `B` is being analyzed and the `assert()` in `f()` happens to call a hypothetical virtual function `v()` defined in `A` and later overridden in `C`? My guess is that it might be a problem because you'd be assuming contract `B` as context and not taking into account state variables from `C`. But maybe that's already handled in some way?

File: libsolidity/formal/Predicate.h (hunk 2 above)
Comment (cameel) [reply libsolidity/formal/Predicate.h:175]: `createSymbolicBlock()` has `nullptr` as a default value for it and I guess there might be calls not visible in the diff that do not set its value explicitly.

File: test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol (hunk 1 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol:18]: If `B` is being analyzed and the inheritance structure is the same as in this example, `C.v()` doesn't exist yet, and a call to `v()` will either be `B.v()` or `A.v()`. When `C` is being analyzed, its `f()` is actually `B.f()`, and a call to `v()` in `B` will correctly call `C.v()`.
tl;dr: the encoding is correct with respect to the calls themselves, the bug was only in how the cex generator saw the function predicates. I'll add a test similar to what you described.

File: libsolidity/formal/Predicate.h (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/Predicate.h:175]: I don't think so, but it will be possible when we have free functions.

Comment (leonardoalt) [reply libsolidity/formal/Predicate.h:175]: Currently they all set it explicitly, but `m_currentContract` might be nullptr when we have free functions (in CHC.cpp)

File: test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol (hunk 1 above)
Comment (cameel) [reply test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_2.sol:18]: > `C.v()` doesn't exist yet, and a call to `v()` will either be `B.v()` or `A.v()`

Oh, you're right. It can't call `C.v()` when we're analyzing `B`.

**Review by cameel — DISMISSED** [review]
LGTM

**Comment by leonardoalt** [comment]
Found a bug when writing more tests so blocking for now

**Comment by leonardoalt** [comment]
@hrkrshnn @cameel this is ready for review again

**Comment by hrkrshnn** [comment]
@leonardoalt What changed here since last time?

**Comment by leonardoalt** [comment]
Nothing, I just brought it back from draft after the bugfix PR that blocked this was merged.

File: test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_1.sol [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+contract C is B {
+	uint y;
+	function g() public {
+		x = 1;
+		f();
+	}
+}
+// ----
+// Warning 6328: (85-99): CHC: Assertion violation happens here.\nCounterexample:\ny = 0, x = 1\n\nTransaction trace:\nC.constructor()\nState: y = 0, x = 0\nC.g()\n    B.f() -- internal call\nState: y = 0, x = 1\nB.f()
```
Comment (hrkrshnn) [inline test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_1.sol:18]: Didn't example like this already work, in the previous PR? What is special here?

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_1.sol:18]: The special part here, which this PR fixes, is when an assertion fails in a base contract and the most derived contract has extra state variables.

Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/inheritance/base_contract_assertion_fail_1.sol:18]: The bug is only related to counterexample generation, not to encoding/solving.

File: libsolidity/formal/Predicate.cpp [hunk 4]
```diff
@@ -130,18 +133,8 @@ FunctionCall const* Predicate::programFunctionCall() const
...
-
-	auto const* node = m_node;
-	while (auto const* scopable = dynamic_cast<Scopable const*>(node))
-	{
-		node = scopable->scope();
-		if (auto const* fun = dynamic_cast<FunctionDefinition const*>(node))
-			return SMTEncoder::stateVariablesIncludingInheritedAndPrivate(*fun);
-	}
+	if (m_contractContext)
+		return SMTEncoder::stateVariablesIncludingInheritedAndPrivate(*m_contractContext);
```
Comment (hrkrshnn) [inline libsolidity/formal/Predicate.cpp:137]: This is mainly what fixes the bug, is it?

Comment (leonardoalt) [reply libsolidity/formal/Predicate.cpp:137]: Yep. But for that we need to always pass the context

**Review by ekpyron — APPROVED** [review]

**Review by hrkrshnn — APPROVED** [review]
